controllers/mr_inferenceservice_controller.go

```diff
@@ -0,0 +1,414 @@
+		return ctrl.Result{}, nil
+	}
+
+	var modelRegistryNamespace string
+	if mrNSFromISVC, ok := isvc.Labels[constants.ModelRegistryNamespaceLabel]; ok {
+		// look for model registry located at specific namespace, as specified in isvc
+		modelRegistryNamespace = mrNSFromISVC
+	} else {
+		// look for model registry inside current namespace
+		modelRegistryNamespace = req.Namespace
+	}
+
+	// setup grpc connection to ml-metadata
+	ctxTimeout, cancel := context.WithTimeout(context.Background(), time.Second*5)
+	defer cancel()
+
+	log.Info("Creating model registry service..")
+	mr, conn, err := r.initModelRegistryService(ctx, ctxTimeout, log, modelRegistryNamespace)
+	if err != nil {
+		log.Error(err, "Stop ModelRegistry InferenceService reconciliation")
+		return ctrl.Result{Requeue: true, RequeueAfter: time.Second * 10}, err
+	} else if mr == nil {
+		// There is no model registry installed, do not requeue
+		log.Info("Cannot find ModelRegistry in given namespace, stopping reconciliation")
+		return ctrl.Result{}, nil
+	}
+
+	if conn != nil {
+		defer conn.Close()
+	}
+
+	// Retrieve or create the ServingEnvironment associated to the current namespace
+	servingEnvironment, err := r.getOrCreateServingEnvironment(log, mr, req.Namespace)
+	if err != nil {
+		return ctrl.Result{Requeue: true}, err
+	}
+
+	// Check if the InferenceService instance is marked to be deleted, which is
+	// indicated by the deletion timestamp being set.
+	isMarkedToBeDeleted := isvc.GetDeletionTimestamp() != nil
+
+	// Let's add a finalizer. Then, we can define some operations which should
+	// occurs before the custom resource to be deleted.
+	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/finalizers
+	if !isMarkedToBeDeleted && !controllerutil.ContainsFinalizer(isvc, modelRegistryFinalizer) {
+		log.Info("Adding Finalizer for ModelRegistry")
+		if ok := controllerutil.AddFinalizer(isvc, modelRegistryFinalizer); !ok {
+			log.Error(err, "Failed to add finalizer into the InferenceService custom resource")
+			return ctrl.Result{Requeue: true}, nil
+		}
+
+		if err = r.client.Update(ctx, isvc); err != nil {
+			log.Error(err, "Failed to update InferenceService custom resource to add finalizer")
+			return ctrl.Result{}, err
+		}
+	}
+
+	var is *openapi.InferenceService
+
+	if okIsId {
+		// Retrieve the IS from model registry using the id
+		log.Info("Retrieving model registry InferenceService by id", "isId", isId)
+		is, err = mr.GetInferenceServiceById(isId)
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("unable to find InferenceService with id %s in model registry: %w", isId, err)
+		}
+	} else if okRegisteredModelId || okModelVersionId {
+		// No corresponding InferenceService in model registry, create new one
+		is, err = r.createMRInferenceService(log, mr, isvc, *servingEnvironment.Id, registeredModelId, modelVersionId)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
```

Comment:
> If it is important to write a record in the MR about that a model got deployed while the MR was down, then all is good.

I'd lean into this use-case, as the purpose of the InferenceService entities in MR is a track record of audit log of how the model "was used". 

My understanding is that the reconcile loop will try until MR comes back up and so a record of InferenceService will then be tracked and marked as "archived".

Reply:
> My understanding is that the reconcile loop will try until MR comes back up and so a record of InferenceService will then be tracked and marked as "archived".

Right, that's what I understand from the code: if the registry was down while the ISVC was both created and deleted, this will retry until the registry is up again and the records in the registry can be created.

The only point that end-users may see weird is that the ISVC will be stuck in `Terminating` state until the registry recovers. But given there is no messaging layer that can help with this, I don't see another way around it.

So, let's go with this as is.